Source/WebCore/dom/PointerEvent.cpp

```diff
@@ -86,6 +96,54 @@ PointerEvent::PointerEvent()
 {
 }
 
+// Calculated in accordance with https://w3c.github.io/pointerevents/#converting-between-tiltx-tilty-and-altitudeangle-azimuthangle.
+static PointerEventAngle angleFromTilt(long tiltX, long tiltY)
+{
+    const double tiltXRadians = tiltX * piDouble / 180;
+    const double tiltYRadians = tiltY * piDouble / 180;
+    double azimuthAngle = 0;
+
+    if (!tiltX && tiltY)
+        azimuthAngle = tiltY > 0 ? piOverTwoDouble : 3 * piOverTwoDouble;
+
+    else if (tiltX && !tiltY)
+        azimuthAngle = tiltX < 0 ? piDouble : azimuthAngle;
+
+    else if (abs(tiltX) == 90 || abs(tiltY) == 90)
+        azimuthAngle = 0;
+
+    else {
+        azimuthAngle = atan2(tan(tiltYRadians), tan(tiltXRadians));
+        azimuthAngle = azimuthAngle < 0 ? azimuthAngle + 2 * piDouble : azimuthAngle;
+    }
+
+    double altitudeAngle = 0;
+
+    if (abs(tiltX) == 90 || abs(tiltY) == 90)
+        altitudeAngle = 0;
+
+    else if (!tiltX)
+        altitudeAngle = piOverTwoDouble - abs(tiltYRadians);
+
+    else if (!tiltY)
+        altitudeAngle = piOverTwoDouble - abs(tiltXRadians);
+
+    else
+        altitudeAngle =  atan(1.0 / sqrt(pow(tan(tiltXRadians), 2) + pow(tan(tiltYRadians), 2)));
+
+    return { altitudeAngle, azimuthAngle };
+}
+
+// This algorithm is equivalent to the one provided in https://w3c.github.io/pointerevents/#converting-between-tiltx-tilty-and-altitudeangle-azimuthangle.
+static PointerEventTilt tiltFromAngle(double altitudeAngle, double azimuthAngle)
+{
+    const double radiansToDegrees = 180 / piDouble;
+    double tiltXRadians = !altitudeAngle ? cos(azimuthAngle) * piOverTwoDouble : atan(cos(azimuthAngle) / tan(altitudeAngle));
```

Comment:
should this ! be an areEssentiallyEqual(altitudeAngle, 0)? Floating point equality is fun.

Reply:
Using areEssentiallyEqual seems to make the relevant pointer event test fail.